runtime/parachains/src/paras_inherent.rs

```diff
@@ -295,15 +519,265 @@ pub mod pallet {
+	for unchecked_bitfield in unchecked_bitfields {
+		ensure2!(
+			unchecked_bitfield.unchecked_payload().0.len() == expected_bits,
+			crate::inclusion::Error::<T>::WrongBitfieldSize,
+			EARLY_RETURN,
+			continue
+		);
+
+		ensure2!(
+			unchecked_bitfield.unchecked_payload().0.clone() & disputed_bitfield.0.clone() ==
+				all_zeros,
+			crate::inclusion::Error::<T>::BitfieldReferencesFreedCore,
+			EARLY_RETURN,
+			continue
+		);
+
+		ensure2!(
+			last_index.map_or(true, |last| last < unchecked_bitfield.unchecked_validator_index()),
+			crate::inclusion::Error::<T>::BitfieldDuplicateOrUnordered,
+			EARLY_RETURN,
+			continue
+		);
+
+		ensure2!(
+			(unchecked_bitfield.unchecked_validator_index().0 as usize) < validators.len(),
+			crate::inclusion::Error::<T>::ValidatorIndexOutOfBounds,
+			EARLY_RETURN,
+			continue
+		);
+
+		let validator_index = unchecked_bitfield.unchecked_validator_index();
+
+		let validator_public = &validators[validator_index.0 as usize];
+
+		// only check the signatures when returning early
+		if EARLY_RETURN {
+			let signed_bitfield = if let Ok(signed_bitfield) =
+				unchecked_bitfield.try_into_checked(&signing_context, validator_public)
+			{
+				signed_bitfield
+			} else {
+				fail!(crate::inclusion::Error::<T>::InvalidBitfieldSignature);
+			};
+			bitfields.push(signed_bitfield.into_unchecked());
+		} else {
+			bitfields.push(unchecked_bitfield);
+		}
+
+		last_index = Some(validator_index);
+	}
+	Ok(bitfields)
+}
+
+/// Filter out any candidates, that have a concluded invalid dispute.
+///
+/// `scheduled` follows the same naming scheme as provided in the
+/// guide: Currently `free` but might become `occupied`.
+/// For the filtering here the relevant part is only the current `free`
+/// state.
+///
+/// `candidate_has_concluded_invalid_dispute` must return `true` if the candidate
+/// is disputed, false otherwise
+fn sanitize_backed_candidates<
+	T: crate::inclusion::Config,
+	F: Fn(CandidateHash) -> bool,
+	const EARLY_RETURN: bool,
+>(
+	relay_parent: T::Hash,
+	mut backed_candidates: Vec<BackedCandidate<T::Hash>>,
+	candidate_has_concluded_invalid_dispute: F,
+	scheduled: &[CoreAssignment],
+) -> Result<Vec<BackedCandidate<T::Hash>>, Error<T>> {
+	let n = backed_candidates.len();
+	// Remove any candidates that were concluded invalid.
+	backed_candidates.retain(|backed_candidate| {
+		!candidate_has_concluded_invalid_dispute(backed_candidate.candidate.hash())
+	});
+	ensure2!(backed_candidates.len() == n, Error::<T>::CandidateConcludedInvalid, EARLY_RETURN);
+
+	// Assure the backed candidate's `ParaId`'s core is free.
+	// This holds under the assumption that `Scheduler::schedule` is called _before_.
+	// Also checks the candidate references the correct relay parent.
+	backed_candidates.retain(|backed_candidate| {
+		let desc = backed_candidate.descriptor();
+		desc.relay_parent == relay_parent &&
+			scheduled.iter().any(|core| core.para_id == desc.para_id)
```

Comment:
Can we make `scheduled` a set? If it just O(scheduled.len()) to collect into the set and then each look up is O(1) should be better than doing a O(backed_candidate.len() * scheduled.len())

Reply:
We would go from `O(backed_candidate.len() * scheduled.len())` to `O(backed_candidate.len() + scheduled.len())`, which is still a fair improvement, since we expect more than two cores and backed candidates.